conf/db.c

```diff
@@ -362,15 +399,41 @@ M0_INTERNAL int m0_confdb_create(struct m0_be_seg      *seg,
 			break;
 		M0_ASSERT(obj_ctx.oc_obj != NULL);
 		/* discard const */
-		key = M0_FID_BUF((struct m0_fid *)m0_conf_objx_fid(obj_ctx.oc_obj));
-		val = M0_BUF_INIT(m0_confx_sizeof() +
-				  sizeof(struct confx_allocator), &obj_ctx);
-		rc = M0_BE_OP_SYNC_RET(op, m0_be_btree_insert(btree, tx,
-							      &op, &key, &val),
-				       bo_u.u_btree.t_rc);
+
+		k_ptr = (struct m0_fid *)m0_conf_objx_fid(obj_ctx.oc_obj);
+		ksize = sizeof (struct m0_fid);
+
+		v_ptr = &obj_ctx;
+		vsize = m0_confx_sizeof() + sizeof(struct confx_allocator);
+
+		rec.r_key.k_data   = M0_BUFVEC_INIT_BUF(&k_ptr, &ksize);
+		rec.r_val          = M0_BUFVEC_INIT_BUF(&v_ptr, &vsize);
+
+		put_cb.c_act       = confd_btree_kv_put_cb;
+		put_cb.c_datum     = &rec;
+		rc = M0_BTREE_OP_SYNC_WITH_RC(&kv_op,
+					      m0_btree_put(&btree, &rec, &put_cb,
+							   &kv_op, tx));
+		if (rc != 0) {
+			/** Delete all objects added to the btree. */
+			while (i--) {
+				if (confx_obj_dup(&alloc, &obj_ctx.oc_obj,
+						  M0_CONFX_AT(conf, i)) != 0)
+					continue;
+
+				k_ptr = (struct m0_fid *)
+					     m0_conf_objx_fid(obj_ctx.oc_obj);
+				M0_BTREE_OP_SYNC_WITH_RC(&kv_op,
+							 m0_btree_del(&btree,
+								      &rec.r_key,
+								      NULL,
+								      &kv_op,
+								      tx));
+			}
+		}
 	}
+	M0_BTREE_OP_SYNC_WITH_RC(&b_op, m0_btree_close(&btree, &b_op));
```

Comment:
shouldn't this be under `rc !=0`

Reply:
No this tree needs to be closed because the btree pointer is not available to the callers of this routine. Other routines which need to operate on this btree will need to re-open the btree.